src/Compilers/Test/Core/SourceGeneration/TestGenerators.cs

```diff
@@ -76,28 +76,47 @@ public SingleFileTestGenerator2(string content, string hintName = "generatedFile
         }
     }
 
-    internal class CallbackGenerator : ISourceGenerator
+    internal class CallbackGenerator(
+        Action<GeneratorInitializationContext> onInit,
+        Action<GeneratorExecutionContext> onExecute,
+        Func<(string hintName, string? source)>? computeSource,
+        Func<(string hintName, SourceText sourceText)>? computeSourceText)
```

Comment:
callbacks so tests can generate source in either string form, or source-text form @chsienki .